### PR TITLE
8286491: Withdraw ValueObject and IdentityObject and introduce value and identity class modifiers

### PR DESCRIPTION
**Pull request by sadayapalam**
Changes to javac, hotspot runtime, core libs and asmtools sources and tests

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issues
 * [JDK-8286491](https://bugs.openjdk.java.net/browse/JDK-8286491): Withdraw ValueObject and IdentityObject and introduce value and identity class modifiers
 * [JDK-8237952](https://bugs.openjdk.java.net/browse/JDK-8237952): [lworld] Examine Class::getAnnotatedInterfaces and other annotation-related API

### Contributors
 * David Simms `<dsimms@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/688/head:pull/688` \
`$ git checkout pull/688`

Update a local copy of the PR: \
`$ git checkout pull/688` \
`$ git pull https://git.openjdk.java.net/valhalla pull/688/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 688`

View PR using the GUI difftool: \
`$ git pr show -t 688`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/688.diff">https://git.openjdk.java.net/valhalla/pull/688.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -907,7 +907,7 @@ void ClassFileParser::parse_interfaces(const ClassFileStream* stream,
         Exceptions::fthrow(
           THREAD_AND_LOCATION,
           vmSymbols::java_lang_IncompatibleClassChangeError(),
-          "Inline type %s attempts to implement interface java.lang.IdentityObject",
+          "Inline type %s has an identity type as supertytype",
```
Comment (Thihup) [inline src/hotspot/share/classfile/classFileParser.cpp:910]: Typo: supertype

**Comment by sadayapalam** [comment]
This PR implements the following:

    - Withdraw the top interface types ValueObject and IdentityObject
    - Add support for identity and value contextual keyword modifiers for flagging value'ness/identityness of a class/interface.
    - Redefine ACC_VALUE to be 0x40

Notes for hotspot, core-libs and other non-langtools changes reviewers:

Changes necessary to rip out PERMITS_VALUE flag bit are not included - I have simply redefined the value of the flag bit to "get it out of the way"

This change set is failing a few tier 1,2,3 tests - I will attach the list of failing tests. If you will provide me with fixes for these and also any additional rectifications/amendments/changes you see fit, I am happy to include them - Thanks!

**Comment by sadayapalam** [comment]
Tier1 Failing Tests:

Hotspot/runtime:

runtime/valhalla/inlinetypes/classfileparser/ACC_CFETest.java
runtime/valhalla/inlinetypes/identityObject/TestIdentityObject.java
runtime/valhalla/inlinetypes/primitiveObject/TestPrimitiveObject.java
runtime/valhalla/inlinetypes/testSupers/TestSuperClasses.java
runtime/valhalla/inlinetypes/verifier/VerifierInlineTypes.java
runtime/valhalla/inlinetypes/withfieldTests/RunWithfieldTests.java
serviceability/dcmd/vm/ClassHierarchyTest.java

Langtools:
tools/javac/ClassFileModifiers/ClassModifiers.java
tools/javac/SynchronizedClass.java
tools/javac/classfiles/attributes/innerclasses/InnerAnnotationsInInnerClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerAnnotationsInInnerEnumTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInAnonymousClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInInnerAnnotationTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInInnerClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInInnerEnumTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInInnerInterfaceTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesInLocalClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerClassesTest.java
tools/javac/classfiles/attributes/innerclasses/InnerEnumInInnerAnnotationTest.java
tools/javac/classfiles/attributes/innerclasses/InnerEnumInInnerEnumTest.java
tools/javac/classfiles/attributes/innerclasses/InnerEnumInInnerInterfaceTest.java
tools/javac/classfiles/attributes/innerclasses/InnerEnumsInInnerClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerInterfacesInInnerClassTest.java
tools/javac/classfiles/attributes/innerclasses/InnerInterfacesInInnerEnumTest.java
tools/javac/diags/CheckExamples.java
tools/javac/platform/createsymbols/CreateSymbolsTest.java
tools/javac/processing/model/completionfailure/NoAbortForBadClassFile.java
tools/javap/TestSuperclass.java

**Comment by sadayapalam** [comment]
Tier2 Failing Tests:

Hotspot/runtime:

runtime/cds/appcds/TestDumpClassListSource.java
runtime/cds/appcds/cacheObject/CheckCachedMirrorTest.java
runtime/cds/appcds/cacheObject/CheckCachedResolvedReferences.java
runtime/cds/appcds/customLoader/ClassListFormatA.java
runtime/cds/appcds/customLoader/ClassListFormatD.java
runtime/cds/appcds/customLoader/ClassListFormatE.java
runtime/cds/appcds/customLoader/CustomClassListDump.java
runtime/cds/appcds/customLoader/HelloCustom.java
runtime/cds/appcds/customLoader/HelloCustom_JFR.java
runtime/cds/appcds/customLoader/LoaderSegregationTest.java
runtime/cds/appcds/customLoader/OldClassAndInf.java
runtime/cds/appcds/customLoader/ParallelTestMultiFP.java
runtime/cds/appcds/customLoader/ParallelTestSingleFP.java
runtime/cds/appcds/customLoader/PrintSharedArchiveAndExit.java
runtime/cds/appcds/customLoader/ProtectionDomain.java
runtime/cds/appcds/customLoader/SameNameInTwoLoadersTest.java
runtime/cds/appcds/customLoader/UnintendedLoadersTest.java
runtime/cds/appcds/customLoader/UnloadUnregisteredLoaderTest.java
runtime/cds/appcds/jvmti/transformRelatedClasses/TransformInterfaceImplementorAppCDS.java
runtime/cds/appcds/jvmti/transformRelatedClasses/TransformSuperSubAppCDS.java
runtime/cds/appcds/loaderConstraints/LoaderConstraintsTest.java#custom-cl


corelibs/jdk:

javax/naming/module/RunBasic.java (Spurious ?)
java/net/ProxySelector/MultiThreadedSystemProxies.java (Spurious ?)

**Comment by MrSimms** [comment]
Patch to remove IdentityObject references from micros (yet to complete testing, compiles for now)...

[remove_idobject_benchmark.zip](https://github.com/openjdk/valhalla/files/8699107/remove_idobject_benchmark.zip)

**Comment by TheShermanTanker** [comment]
Bit of a weird question, but considering `SomeObject.val` might end up as a language feature in the future (https://mail.openjdk.java.net/pipermail/valhalla-spec-observers/2022-May/001934.html) and that we already have the ref keyword reserved, would it be worth considering reusing those keywords for modifying the "identitiness" of an object instead of having an extra keyword instead?

**Comment by forax** [comment]
Foo.ref is a type projection, it does not change the behavior of the class at runtime.
So X.ref (or X.val) is only a valid operation on a value class, a class which at runtime will not use its address when == is called by example.

String.ref or String.val does not exist, otherwise it means that == should have different behaviors at runtime depending on the type that store the string.

**Comment by MrSimms** [comment]
> Bit of a weird question, but considering `SomeObject.val` might end up as a language feature in the future (https://mail.openjdk.java.net/pipermail/valhalla-spec-observers/2022-May/001934.html) and that we already have the ref keyword reserved, would it be worth considering reusing those keywords for modifying the "identitiness" of an object instead of having an extra keyword?

"valhalla-spec-observers" mailing list would be a more appropriate forum.

**Comment by MrSimms** [comment]
Adjust ACC_CFETest (doesn't remove ACC_PERMITS_VALUE, that's follow work)

[ACC_CFETest.zip](https://github.com/openjdk/valhalla/files/8700971/ACC_CFETest.zip)

**Comment by sadayapalam** [comment]
The best way to take forward this PR was discussed elsewhere and it was agreed that given the number of failing tests is small, we can go ahead and integrate this, Each component will independently review the changes, make suitable additional changes and fix the remaining failures. 

Proceeding to integrate on that basis.

**Pull request closed without merging**